src/lib/y2users/autoyast/hash_reader.rb

```diff
@@ -0,0 +1,103 @@
+# Copyright (c) [2021] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2users/parsers/shadow"
+require "y2users/user"
+require "y2users/group"
+require "y2users/password"
+
+module Y2Users
+  module Autoyast
+    # Reader that fills config with hash from Users.Export
+    class HashReader
+      include Yast::Logger
+
+      def initialize(content)
+        @users_export = content
+      end
+
+      def read_to(config)
+        config.attach(read_users + read_groups)
+      end
+
+    private
+
+      attr_reader :users_export
+
+      def read_users
+        exported_users = users_export["users"]
+        exported_users.map do |e_user|
+          res = User.new(e_user["username"])
+          res.gecos = [e_user["fullname"]]
+          res.gid = e_user["gid"]
+          res.home = e_user["home"]
+          res.shell = e_user["shell"]
+          res.uid = e_user["uid"]
+          res.password = create_password(e_user)
+          res
+        end
+      end
+
+      def read_groups
+        exported_groups = users_export["groups"]
+        exported_groups.map do |e_group|
+          res = Group.new(e_group["groupname"])
+          res.gid = e_group["gid"]
+          users_name = e_group["userlist"].split(",")
+          res.users_name = users_name
+          res.source = :local
+          res
+        end
+      end
+
+      # shadow attrs without password and username which is done manually
+      SORTED_SHADOW_ATTRS = [
```

Comment:
I see some duplication here with the `UsersSimple::Reader`. Both have almost the same code for creating the passwords. Maybe we could generalize it.

Reply:
sadly not. As you can see here is a big inconsistency between attrs names see e.g. min, max..and UsersSimple use ShadowMin and ShadowMax and it is not in nested hash.

Reply:
Ok, I thought it was only a matter of key names, but if the structure is different ...

Reply:
yes, see https://github.com/yast/yast-users/pull/272/files/f60488738f5aac75e34b295676e45e8405b8ad5f#diff-58cefa9c64c06ad09607ab955a8b672cc20183a654199e76fd3f6ce784c04330R94 it reads that nested hash. I think it is reason why UsersSimple has shadow prefix everywhere to distinguish between shadow and passwd keys.